File: adwin/src/main/java/com/bharani/adwin/utils/ADBuckets.java
package com.bharani.adwin.utils;

public class ADBuckets {

    private ADBucket[] buckets;
    private ADBuckets prev;
    private ADBuckets next;
    private int count;
    private double sum;

    //Initialize all buckets
    //Initialize all prev and next to do the looping
    ADBuckets(int size, ADBuckets prev, ADBuckets next) {

        buckets = new ADBucket[size];
        for (int i = 0; i < size; i++) {
            buckets[i] = new ADBucket(0, 0);
        }

        if (prev != null) {
            prev.setNext(this);
            this.setPrev(prev);
        }
        if (next != null) {
            this.setNext(next);
            next.setPrev(this);
        }
        count = 0;
    }

    public int getCount() {
        return count;
    }

    public void setCount(int count) {
        this.count = count;
    }

    public ADBucket[] getBuckets() {
        return buckets;
    }

    public ADBuckets getPrev() {
        return prev;
    }

    public void setPrev(ADBuckets prev) {
        this.prev = prev;
    }

    public ADBuckets getNext() {
        return next;
    }

    public void setNext(ADBuckets next) {
        this.next = next;
    }

    public double getSum() {
        return sum;
    }

    public void setSum(double sum) {
        this.sum = sum;
    }

    public void setBuckets(ADBucket[] buckets) {
        this.buckets = buckets;
    }
}

File: algorithm/ADBuckets.py

class ADBuckets:

    def __init__(self,size=6,prev=None,next=None):
        self.size = size
        self.buckets = [None]*self.size
        self.prev = None
        self.next = None
        self.count=0
        self.sum =0.0

        for i in range(size):
            self.buckets[i] = ADBucket(0,0)


        if prev is not None:
            self.prev = prev
            prev.next = self

        if next is not None:
            self.next = next
            next.prev = self
